README.md
# VIOOH-tech-test

## Problem
Given datasets of songs played and user-data, determine the top 10 songs played in the top 50 sessions
in the dataset, where a session is defined to be a collection of songs played by a given user where the interval between each element is not greater than 20 minutes.

## Usage

Clone the repo with ```git clone https://github.com/MagusMachinae/viooh-tech-test```. All
commands to be executed are provided at the top of the comment block in ```viooh.core```

Due to storage limits on git-hub, it is necessary to download the data set from
http://ocelma.net/MusicRecommendationDataset/lastfm-1K.html
, placing it in ```resources/```, it is useful at this stage to rename the file containing the data of songs played to make the next step less tedious, the readme and code in the namespace assumes it has been renamed to ```song-data.tsv```.

Create resources sub-directories required for generated files by calling the ```(make-parents ...)``` form in the comment expression.

Call ```clojure (data->edn! "resources/song-data.tsv")``` to generate a set of files containing hash-maps of data representing plays of songs sorted by user.

Call ```clojure (generate-user-sessions! user-vector)``` to output a list of files containing session data with a structure of ```{:session-name user-session-# :play-count int :session-data [play & plays]}```

Call ```(calculate-answer user-vector)``` to get the result, or evaluate the whole ```(answer->tsv! ...)``` expression to print the calculated result to a .tsv file.

## Notes and Commentary
While I had a strong intuition about how to go about solving the problem, I had difficulty concretely visualising each step, so I decided to start with focusing on IO and data parsing, which would also help me persist the results of each step and get data back that I could use to help reason about what data structures my functions needed to operate on by generating output files.

Because of the nature of the problem, I decided to approach it more interactively, grabbing small sections of data to work on in the comment block, and gradually building up the logic to calculate the final count and testing against generated data, naming processes once I had reached a clear boundary at which to mark progress, that gave me a better vantage point to the rest of the solution and improve upon it. For example, generating the files told me that only 647 users are represented in the dataset, so I could adjust my functions for pulling the users out of the source file to limit the result-set. Seperating the output into files also helped the process of collecting them into sessions, as pre-sorting the files meant that a test on users would be redundant. Once I had a good understanding of the data, about the time I spat out ```top50.edn```, I then could focus on the transformation pipeline fully.

I consistently ran into heap size issues and long calculation times during the second stage of computation involving running over the sessions, so I had to adapt the logic of my implementation to work with clojure's reducers to take advantage of the more efficient reduction algorithms.

When it came to choosing my libraries, tick seemed an obvious choice because of its interval calculus for computing times, and I needed a .csv parsing library to handle reading and writing .tsv, but aside from that, I felt that the core language features and tactical application of hash-maps would be more than enough to provide an easy solution. Given more time, I would rewrite the initial portions of the transformation to avoid reading to files, even though it helped me avoid reasoning about nested loops. Additionally, I would take the time to look at how I can better use reducers and multi-threading do decrease computation time.



## License

Copyright © 2021 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-01-22
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-01-22
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/viooh/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/viooh/compare/0.1.0...0.1.1


test/viooh/core_test.clj
(ns viooh.core-test
  (:require [clojure.test :refer :all]
            [viooh.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/viooh/data.clj
(ns viooh.data)

(def user-vector ["user_000001" "user_000002" "user_000003" "user_000004" "user_000005" "user_000006" "user_000007" "user_000008" "user_000009" "user_000010" "user_000011" "user_000012" "user_000013" "user_000014" "user_000015" "user_000016" "user_000017" "user_000018" "user_000019" "user_000020" "user_000021" "user_000022" "user_000023" "user_000024" "user_000025" "user_000026" "user_000027" "user_000028" "user_000029" "user_000030" "user_000031" "user_000032" "user_000033" "user_000034" "user_000035" "user_000036" "user_000037" "user_000038" "user_000039" "user_000040" "user_000041" "user_000042" "user_000043" "user_000044" "user_000045" "user_000046" "user_000047" "user_000048" "user_000049" "user_000050" "user_000051" "user_000052" "user_000053" "user_000054" "user_000055" "user_000056" "user_000057" "user_000058" "user_000059" "user_000060" "user_000061" "user_000062" "user_000063" "user_000064" "user_000065" "user_000066" "user_000067" "user_000068" "user_000069" "user_000070" "user_000071" "user_000072" "user_000073" "user_000074" "user_000075" "user_000076" "user_000077" "user_000078" "user_000079" "user_000080" "user_000081" "user_000082" "user_000083" "user_000084" "user_000085" "user_000086" "user_000087" "user_000088" "user_000089" "user_000090" "user_000091" "user_000092" "user_000093" "user_000094" "user_000095" "user_000096" "user_000097" "user_000098" "user_000099" "user_000100" "user_000101" "user_000102" "user_000103" "user_000104" "user_000105" "user_000106" "user_000107" "user_000108" "user_000109" "user_000110" "user_000111" "user_000112" "user_000113" "user_000114" "user_000115" "user_000116" "user_000117" "user_000118" "user_000119" "user_000120" "user_000121" "user_000122" "user_000123" "user_000124" "user_000125" "user_000126" "user_000127" "user_000128" "user_000129" "user_000130" "user_000131" "user_000132" "user_000133" "user_000134" "user_000135" "user_000136" "user_000137" "user_000138" "user_000139" "user_000140" "user_000141" "user_000142" "user_000143" "user_000144" "user_000145" "user_000146" "user_000147" "user_000148" "user_000149" "user_000150" "user_000151" "user_000152" "user_000153" "user_000154" "user_000155" "user_000156" "user_000157" "user_000158" "user_000159" "user_000160" "user_000161" "user_000162" "user_000163" "user_000164" "user_000165" "user_000166" "user_000167" "user_000168" "user_000169" "user_000170" "user_000171" "user_000172" "user_000173" "user_000174" "user_000175" "user_000176" "user_000177" "user_000178" "user_000179" "user_000180" "user_000181" "user_000182" "user_000183" "user_000184" "user_000185" "user_000186" "user_000187" "user_000188" "user_000189" "user_000190" "user_000191" "user_000192" "user_000193" "user_000194" "user_000195" "user_000196" "user_000197" "user_000198" "user_000199" "user_000200" "user_000201" "user_000202" "user_000203" "user_000204" "user_000205" "user_000206" "user_000207" "user_000208" "user_000209" "user_000210" "user_000211" "user_000212" "user_000213" "user_000214" "user_000215" "user_000216" "user_000217" "user_000218" "user_000219" "user_000220" "user_000221" "user_000222" "user_000223" "user_000224" "user_000225" "user_000226" "user_000227" "user_000228" "user_000229" "user_000230" "user_000231" "user_000232" "user_000233" "user_000234" "user_000235" "user_000236" "user_000237" "user_000238" "user_000239" "user_000240" "user_000241" "user_000242" "user_000243" "user_000244" "user_000245" "user_000246" "user_000247" "user_000248" "user_000249" "user_000250" "user_000251" "user_000252" "user_000253" "user_000254" "user_000255" "user_000256" "user_000257" "user_000258" "user_000259" "user_000260" "user_000261" "user_000262" "user_000263" "user_000264" "user_000265" "user_000266" "user_000267" "user_000268" "user_000269" "user_000270" "user_000271" "user_000272" "user_000273" "user_000274" "user_000275" "user_000276" "user_000277" "user_000278" "user_000279" "user_000280" "user_000281" "user_000282" "user_000284" "user_000285" "user_000286" "user_000288" "user_000289" "user_000290" "user_000291" "user_000292" "user_000293" "user_000294" "user_000295" "user_000296" "user_000297" "user_000298" "user_000299" "user_000300" "user_000301" "user_000302" "user_000303" "user_000304" "user_000305" "user_000306" "user_000307" "user_000308" "user_000309" "user_000310" "user_000311" "user_000312" "user_000313" "user_000314" "user_000315" "user_000316" "user_000317" "user_000318" "user_000319" "user_000320" "user_000321" "user_000322" "user_000323" "user_000324" "user_000325" "user_000326" "user_000327" "user_000328" "user_000329" "user_000330" "user_000331" "user_000332" "user_000333" "user_000334" "user_000335" "user_000336" "user_000337" "user_000338" "user_000339" "user_000340" "user_000341" "user_000342" "user_000343" "user_000344" "user_000345" "user_000346" "user_000347" "user_000348" "user_000349" "user_000350" "user_000351" "user_000352" "user_000353" "user_000354" "user_000355" "user_000356" "user_000357" "user_000358" "user_000359" "user_000360" "user_000361" "user_000362" "user_000363" "user_000364" "user_000366" "user_000367" "user_000368" "user_000369" "user_000370" "user_000371" "user_000372" "user_000373" "user_000374" "user_000375" "user_000376" "user_000377" "user_000378" "user_000379" "user_000380" "user_000381" "user_000382" "user_000383" "user_000384" "user_000385" "user_000386" "user_000387" "user_000388" "user_000389" "user_000390" "user_000391" "user_000392" "user_000393" "user_000394" "user_000395" "user_000396" "user_000397" "user_000398" "user_000399" "user_000400" "user_000401" "user_000402" "user_000403" "user_000404" "user_000405" "user_000406" "user_000407" "user_000408" "user_000409" "user_000410" "user_000411" "user_000412" "user_000413" "user_000414" "user_000415" "user_000416" "user_000417" "user_000418" "user_000419" "user_000420" "user_000421" "user_000422" "user_000423" "user_000424" "user_000425" "user_000426" "user_000427" "user_000429" "user_000430" "user_000431" "user_000432" "user_000433" "user_000434" "user_000435" "user_000436" "user_000437" "user_000438" "user_000439" "user_000440" "user_000441" "user_000442" "user_000443" "user_000444" "user_000445" "user_000446" "user_000447" "user_000448" "user_000449" "user_000450" "user_000451" "user_000452" "user_000453" "user_000454" "user_000455" "user_000456" "user_000457" "user_000458" "user_000459" "user_000460" "user_000461" "user_000462" "user_000463" "user_000464" "user_000465" "user_000467" "user_000468" "user_000469" "user_000471" "user_000472" "user_000473" "user_000474" "user_000475" "user_000476" "user_000477" "user_000478" "user_000479" "user_000480" "user_000481" "user_000482" "user_000483" "user_000484" "user_000485" "user_000486" "user_000487" "user_000488" "user_000489" "user_000490" "user_000491" "user_000492" "user_000493" "user_000494" "user_000495" "user_000496" "user_000497" "user_000498" "user_000499" "user_000500" "user_000501" "user_000502" "user_000503" "user_000504" "user_000505" "user_000506" "user_000507" "user_000508" "user_000509" "user_000510" "user_000511" "user_000512" "user_000513" "user_000514" "user_000515" "user_000516" "user_000517" "user_000518" "user_000519" "user_000520" "user_000521" "user_000522" "user_000523" "user_000524" "user_000525" "user_000526" "user_000527" "user_000528" "user_000529" "user_000530" "user_000531" "user_000532" "user_000533" "user_000534" "user_000535" "user_000536" "user_000537" "user_000538" "user_000539" "user_000540" "user_000541" "user_000542" "user_000543" "user_000544" "user_000545" "user_000546" "user_000547" "user_000548" "user_000549" "user_000550" "user_000551" "user_000552" "user_000553" "user_000554" "user_000555" "user_000556" "user_000557" "user_000558" "user_000559" "user_000560" "user_000561" "user_000562" "user_000563" "user_000564" "user_000565" "user_000566" "user_000567" "user_000568" "user_000569" "user_000570" "user_000571" "user_000572" "user_000573" "user_000574" "user_000575" "user_000576" "user_000577" "user_000578" "user_000579" "user_000580" "user_000581" "user_000582" "user_000583" "user_000584" "user_000585" "user_000586" "user_000587" "user_000588" "user_000589" "user_000590" "user_000591" "user_000592" "user_000593" "user_000594" "user_000595" "user_000596" "user_000597" "user_000598" "user_000599" "user_000600" "user_000601" "user_000602" "user_000603" "user_000604" "user_000605" "user_000606" "user_000607" "user_000608" "user_000609" "user_000610" "user_000611" "user_000612" "user_000613" "user_000614" "user_000615" "user_000616" "user_000617" "user_000618" "user_000619" "user_000620" "user_000621" "user_000622" "user_000623" "user_000624" "user_000625" "user_000626" "user_000627" "user_000628" "user_000629" "user_000630" "user_000631" "user_000632" "user_000633" "user_000634" "user_000635" "user_000636" "user_000637" "user_000638" "user_000639" "user_000640" "user_000641" "user_000642" "user_000643" "user_000644" "user_000645" "user_000646" "user_000647" "user_000648" "user_000649" "user_000650" "user_000651" "user_000652" "user_000653"])


src/viooh/core.clj
(ns viooh.core
  (:require [clojure-csv.core :as csv]
            [clojure.string :as str]
            [clojure.java.io :as io]
            [tick.alpha.api :as tick]
            [clojure.edn :as edn]
            [clojure.core.reducers :as r]))

(def str-test  "user_000639 \t 2009-04-08T01:57:47Z \t MBID \t The Dogs D'Amour \t MBID \t Fall in Love Again?
      user_000639 \t 2009-04-08T01:53:56Z \t MBID \t The Dogs D'Amour \t MBID \t Wait Until I'm Dead")

(def keys-vector ['user 'time-stamp 'artist-id 'artst-name 'track-id 'track-name])

(def output-folders ["resources/sessions/foo.txt" "resources/raw-hash-maps/foo.txt"])

(def answer '(["Heartless" 1297] ["Love Lockdown" 1295] ["Pinocchio Story (Freestyle Live From Singapore)" 1292] ["Say You Will" 1291] ["Welcome To Heartbreak (Feat. Kid Cudi)" 1291] ["See You In My Nightmares" 1291] ["Amazing (Feat. Young Jeezy)" 1289] ["Paranoid (Feat. Mr. Hudson)" 1289] ["Coldest Winter" 1285] ["Bad News" 1274]))

(defn data->edn! [src]
  "pulls tsv data into files of hash-maps by user"
  (with-open [reader (io/reader src)]
    (doseq [data (drop 1 (csv/parse-csv reader :delimiter \tab))]
      (let [hash-data (zipmap (map keyword keys-vector) (map str/trim data))]
        (spit (str "resources/raw-hash-maps/" (:user hash-data) ".edn")
              (str (dissoc hash-data :artist-id :track-id) "\n")
              :append true)))))

(def user-vector
  (with-open [reader (io/reader "resources/userid-profile.tsv")]
     (into [] (for [x (take 647 (drop 1 (csv/parse-csv reader :delimiter \tab)))]
                   (first (map str/trim x))))))

(defn same-session?
  "Checks if time-1 is within 20 minutes after time-2"
  [time-1 time-2]
  (if (some? time-2)
    (tick/<= (tick/instant time-1)
             (tick/+ (tick/instant time-2)
                     (tick/new-duration 20 :minutes)))
    false))

(same-session? "2009-04-08T01:57:47Z" "2009-04-08T01:37:47Z")

(defn session-stepper
  "steps through the data from a file, collecting it up into session until it reaches
   a boundary, returning the session and the remainder of the lazy-seq of data"
  [file-stream]
  (loop [play-1 (first file-stream)
         play-2  (first (rest file-stream))
         acc     [play-1]
         remainder (rest (rest file-stream))]
    (if (same-session? (:time-stamp play-1) (:time-stamp play-2));;;
      (recur play-2
        (first remainder)
        (conj acc play-2)
        (rest remainder))
      [acc remainder])))

(defn into-sessions!
  "sorts the hashmaps in a file into sessions"
  [file]
  (with-open [reader (io/reader file)]
    (let [file-stream (map read-string (line-seq reader))]
      (loop [session-num 1
             remainder  (second (session-stepper file-stream))
             session-data (first (session-stepper file-stream))]
        (if (some? (first remainder))
          (do (spit (str "resources/sessions/" (:user (first session-data)) "_sessions.edn")
                    (str {:session-name (str (:user (first session-data)) "session-" session-num)
                          :play-count (count session-data)
                          :session-data session-data} "\n")
                    :append true)
            (recur (inc session-num)
                   (second (session-stepper remainder))
                   (first (session-stepper remainder))))
          (spit (str "resources/sessions/" (:user (first session-data)) "_sessions.edn")
                (str {:session-name (str (:user (first session-data)) "session-" session-num)
                      :play-count (count session-data)
                      :session-data session-data} "\n")
                :append true))))))

(defn generate-user-sessions!
  [user-vector]
  (pmap (fn [user] (into-sessions! (str "resources/raw-hash-maps/" user ".edn"))) user-vector))

(defn collect-session
  "concatenates sessions from file"
  [user-vector]
  (apply concat
         (pmap (fn [user] (->> (str "resources/sessions/" user "_sessions.edn")
                               (io/reader)
                               (line-seq)
                               (map read-string)))
               user-vector)))

(defn count-occurences
  "returns a map of songs mapped to the total number number of times they are played"
  [coll]
  (r/fold
    (r/monoid #(merge-with + %1 %2) (constantly {}))
    (fn [m [k cnt]] (assoc m k (+ cnt (get m k 0))))
    (r/map #(vector % 1) (into [] coll))))

(count-occurences   ["foo" "foo" "bar"])

(defn top-by [n k coll]
     (reduce
      (fn [top x]
        (let [top (conj top x)]
          (if (> (count top) n)
            (disj top (first top))
            top)))
      (sorted-set-by #(< (k %1) (k %2))) coll))

(defn top-by-f [n k coll]
     (r/fold
      (fn [top x]
        (let [top (conj top x)]
          (if (> (count top) n)
            (disj top (first top))
            top)))
      (sorted-set-by #(< (k %1) (k %2))) coll))

(defn concat-tracks
  "gets all tracks from the sessions supplied in coll as a single coll."
  [coll]
  (->> coll
   (pmap (fn [session] (get session :session-data)))
   (apply concat)
   (pmap (fn [session-data] (get session-data :track-name)))))

(defn top-ten-tracks
  "sorts the occurences by value"
  [coll]
  (take 10 (sort-by second > (count-occurences (concat-tracks coll)))))

(defn calculate-answer
  "calls the necessary functions when given the user-vector to calculate the result"
  [coll]
  (top-ten-tracks (top-by 50 :play-count (collect-session coll))))

(defn prune-answers
  "prepares answer for printingto tsv."
  [coll]
  (map (fn [[k v]]  [k (str "Plays:" (str v))]) coll))

(defn answer->tsv!
  "takes a coll of the top ten answers and writes it out to a tsv file"
  [coll]
  (spit "answer.tsv" (csv/write-csv (prune-answers coll) :delimiter \tab)))

(comment
 (data->edn! "resources/song-data.tsv")
 (map io/make-parents output-folders)
 (generate-user-sessions! user-vector)
 (answer-tsv! (calculate-answer user-vector))

 (answer->tsv! answer)

 (into-sessions! "resources/raw-hash-maps/user_000001.edn")

 (pmap keyword user-vector)
 (slurp "top50.edn")
 (map read-string (line-seq (io/reader "resources/raw-hash-maps/user_000001.edn")))
 (map read-string
   (line-seq (io/reader (str "resources/sessions/"
                             user
                             "_sessions.edn"))))

 (spit "top50.edn "(top-by 50 :play-count (apply concat
                                            (pmap (fn [user] (map read-string
                                                                  (line-seq (io/reader (str "resources/sessions/"
                                                                                            user
                                                                                            "_sessions.edn"))))) user-vector))))
 (time (top-by 50 :play-count (apply concat
                                            (pmap (fn [user] (map read-string
                                                                  (line-seq (io/reader (str "resources/sessions/"
                                                                                            user
                                                                                            "_sessions.edn"))))) user-vector))))

 (top-by-f 50 :play-count (into [] (apply concat
                                     (pmap (fn [user] (map read-string
                                                           (line-seq (io/reader (str "resources/sessions/"
                                                                                 user
                                                                                 "_sessions.edn")))))
                                           user-vector))))

 (concat-tracks (read-string (slurp "top50.edn")))



 (take 10 (sort-by second > (count-occurences (->> (read-string (slurp "top50.edn"))
                                                   (pmap (fn [session] (get session :session-data)))
                                                   (apply concat)
                                                   (pmap (fn [session-data] (get session-data :track-name)))))))

 (spit "foo.edn"  (apply concat
                    (pmap (fn [user] (->> (str "resources/sessions/" user "_sessions.edn")
                                          (io/reader)
                                          (line-seq)
                                          (map read-string)))
                          user-vector)))

 (for [x (csv/parse-csv str-test :delimiter \tab)]
   ((fn [coll] (dissoc coll :artist-id :track-id))
    (zipmap (map keyword keys-vector) (map str/trim x))))

 (session-stepper (for [x (csv/parse-csv str-test :delimiter \tab)]
                    (dissoc (zipmap (map keyword keys-vector)
                                    (map str/trim x))
                            :artist-id :track-id)))

 (first (first (session-stepper (map read-string (line-seq (io/reader "resources/raw-hash-maps/user_000001.edn"))))))
 (let [file-stream (map read-string (line-seq (io/reader "resources/raw-hash-maps/user_000001.edn")))]
   (loop [
          session-num 1

          remainder  (second (session-stepper file-stream))
          session-data (first (session-stepper file-stream))]

     (if (some? (first remainder))
       (do (spit (str "resources/sessions/" (:user  (first session-data)) "_sessions.edn")
                 (str {:session-name (str (:user (first session-data)) "session-" session-num)
                       :play-count (count session-data)
                       :session-data session-data} "\n")
                 :append true)
         (recur (inc session-num)
           (second (session-stepper remainder))
           (first (session-stepper remainder))))
       (spit (str "resources/sessions/" (:user (first session-data)) "_sessions.edn")
        (str {:session-name (str (:user (first session-data)) "session-" session-num)
              :play-count (count session-data)
              :session-data session-data} "\n")
        :append true))))
 (zipmap (map keyword keys-vector) '("user_000639"
                                     "2009-04-08T01:57:47Z"
                                     "MBID"
                                     "The Dogs D'Amour"
                                     "MBID"
                                     "Fall in Love Again?"))
 (tick/+
  (tick/instant "2009-04-08T01:57:47Z"))
 (some? (first [nil]))
 (spit "target/foo.edn" '("foo" "bar" "baz")))
